### PR TITLE
Encrypted Pools

### PR DESCRIPTION
**Pull request by dwlehman**
This seems fairly complete ~~, but the tests are woefully inadequate~~ . This is on top of #90 and #100. The relevant work begins with a3293e9.

Edit: tests are reasonable now, although they don't check cipher, key size, or luks version.

### CODE REVIEW

**Comment by dwlehman** [comment]
The obvious things missing from testing:

1. verification of absence/presence of LUKS layer as appropriate
2. correct application of various parameters insofar as that is possible

File: library/blivet.py [hunk 1]
```diff
@@ -145,15 +148,72 @@ class BlivetAnsibleError(Exception):
     pass
 
 
-class BlivetVolume(object):
+class BlivetBase(object):
```
Comment (richm) [inline library/blivet.py:151]: do you need unit tests for these changes?  if so, they can be added later after the 15th

File: tests/tests_swap.yml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        name: storage
+      vars:
+        storage_volumes:
+          - name: test1
+            type: disk
+            mount_point: "{{ mount_location }}"
+            fs_type: ext3
+            disks: "{{ unused_disks }}"
+
+    - include_tasks: verify-role-results.yml
```
Comment (richm) [inline tests/tests_swap.yml:54]: Does `verify-role-results.yml` check for idempotence?  Does it do something like `assert: that: some_register_var is not changed`?

Comment (dwlehman) [reply tests/tests_swap.yml:54]: No, the only tests we have for idempotence are explicitly for that purpose -- basically running the role a second time and checking that things are the same afterwards. I don't think I did any of those for this changeset, though. Maybe I'll add one in the morning.

**Comment by pcahyna** [comment]
@sergio-correia the LUKS parameters are needlessly different from the ones used in your role. https://github.com/linux-system-roles/nbde_client/pull/3/files#diff-7eeda618087b49ae876084ab6c73fdbbR9 uses `pass` and `keyfile`. Here one uses `encryption_passphrase` and `encryption_key_file`. Those names should be consistent. (I don't care which one is chosen.)

File: README.md [hunk 3]
```diff
@@ -39,6 +39,27 @@ Accepted values are: `linear`, `striped`, `raid0`, `raid1`, `raid4`, `raid5`, `r
...
+This specifies whether or not the pool will be encrypted using LUKS.
+__WARNING__: Toggling encryption for a pool is a destructive operation, meaning
+             the pool itself will be removed as part of the process of
+             adding/removing the encryption layer.
+
+##### `encryption_passphrase`
+This string specifies a passphrase used to unlock/open the LUKS volume(s).
+
+##### `encryption_key_file`
+This string specifies the full path to the key file used to unlock the LUKS volume(s).
```
Comment (pcahyna) [inline README.md:52]: path on the control node or on the managed host?

Comment (dwlehman) [reply README.md:52]: Everything is on the managed host AFAIK. Any operations to get the key file to the managed nodes is on the user. Maybe I should note that.

Comment (sergio-correia) [reply README.md:52]: > @sergio-correia the LUKS parameters are needlessly different from the ones used in your role. https://github.com/linux-system-roles/nbde_client/pull/3/files#diff-7eeda618087b49ae876084ab6c73fdbbR9 uses `pass` and `keyfile`. Here one uses `encryption_passphrase` and `encryption_key_file`. Those names should be consistent. (I don't care which one is chosen.)

I will adopt `encryption_passphrase` and `encryption_key_file`. These seem more explicit.

Comment (pcahyna) [reply README.md:52]: I think you an use just `passphrase` and `keyfile`. In your role it is obvious that it is encryption related (it is an encryption role). I was concerned about the gratuitous `pass` vs `passphrase` difference or `keyfile` vs. `key_file`.

**Comment by dwlehman** [comment]
@pcahyna @sergio-correia we use the prefixes for a sort of namespacing, so all the encryption-specific parameters have the 'encryption_' prefix. Same for raid, fs, mount, etc. Not sure if there's a benefit for the clevis/tang roles to do something along the same lines.

**Comment by sergio-correia** [comment]
> @pcahyna @sergio-correia we use the prefixes for a sort of namespacing, so all the encryption-specific parameters have the 'encryption_' prefix. Same for raid, fs, mount, etc. Not sure if there's a benefit for the clevis/tang roles to do something along the same lines.

Right, I wont use the prefix. @pcahyna explained he was more concerned about pass/passphrase (we prefer passphrase, so I will update there) and keyfile/key_file

File: README.md (hunk 3 above)
Comment (pcahyna) [reply README.md:52]: @dwlehman would it make sense to have `key_file` on the control node? `nbde_client` uses a keyfile on the control node: https://github.com/linux-system-roles/nbde_client/pull/3#discussion_r439006359 . Are the use cases different enough to justify this difference in semantics?

Comment (dwlehman) [reply README.md:52]: I agree that it is more convenient for users and the consistency between roles is good.

Where on the managed node should we place the key files?

Comment (pcahyna) [reply README.md:52]: > Where on the managed node should we place the key files?

@sergio-correia what do you do in nbde_client?

Comment (sergio-correia) [reply README.md:52]: I define an internal variable `__nbde_client_managed_dir` and have a task to copy the key files there and do the cleanup later. I pass this variable also to the python module, so it knows where to find the key files. It's currently defined as `/var/run/linux-system-roles.nbde_client`

Comment (dwlehman) [reply README.md:52]: For our purposes I think we need to keep the key file around for reboots, while I (think that I) see that you are removing them in the nbde_client role after setting up nbde. For us to specify that they be on the controller necessitates developing a scheme for storing them on the managed nodes.

Comment (sergio-correia) [reply README.md:52]: Yeah, for us it's simpler that we need it to do some operations, then we don't need it anymore and can remove it.

**Pull request merged**